pkg/binary/envoy/debug/pii_test.go

```diff
@@ -0,0 +1,102 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package debug
+
+import (
+	"testing"
+
+	"bitbucket.org/creachadair/shell"
+	"github.com/google/go-cmp/cmp"
+)
+
+func TestProcess(t *testing.T) {
+	var FORMAT_STRING string = "[%START_TIME%] \"%REQ(:METHOD)% %REQ(X-ENVOY-ORIGINAL-PATH?:PATH)% %PROTOCOL%\" %RESPONSE_CODE% %RESPONSE_FLAGS% \"%DYNAMIC_METADATA(istio.mixer:status)%\" \"%UPSTREAM_TRANSPORT_FAILURE_REASON%\" %BYTES_RECEIVED% %BYTES_SENT% %DURATION% %RESP(X-ENVOY-UPSTREAM-SERVICE-TIME)% \"%REQ(X-FORWARDED-FOR)%\" \"%REQ(USER-AGENT)%\" \"%REQ(X-REQUEST-ID)%\" \"%REQ(:AUTHORITY)%\" \"%UPSTREAM_HOST%\" %UPSTREAM_CLUSTER% %UPSTREAM_LOCAL_ADDRESS% %DOWNSTREAM_LOCAL_ADDRESS% %DOWNSTREAM_REMOTE_ADDRESS% %REQUESTED_SERVER_NAME%\n"
+
+	FORMATS, ok := shell.Split(FORMAT_STRING)
+	if !ok {
+		t.Fatalf("error in splitting the FORMAT_STRING: %s", FORMAT_STRING)
+	}
+
+	var VALID_LOG string = `[2019-09-05T17:53:36.908Z] "GET /productpage HTTP/1.1" 200 - "-" "-" 0 5183 47 46 "10.28.2.1" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/76.0.3809.132 Safari/537.36" "4273c09c-2534-962a-b7da-04a04cf39e88" "35.199.144.198" "127.0.0.1:9080" inbound|9080|http|productpage.default.svc.cluster.local - 10.28.2.27:9080 10.28.2.1:0 outbound_.9080_._.productpage.default.svc.cluster.local`
+	var LOG_WITH_ONE_FIELD string = `[2019-09-05T17:53:36.908Z]`
+
+	type Out struct {
+		logs []string
+		err  error
+	}
+	// test cases wrapped up in a map where key is the test case name and value is the struct
+	// defining the inputs and expecte outcome
+	tests := map[string]struct {
+		logs        []string
+		format      string
+		containsPII map[string]bool
+		want        Out
+	}{
+		// positive test cases
+		// only return an array containing one valid log with [%START_TIME%] field hashed
+		"simple": {logs: []string{VALID_LOG, LOG_WITH_ONE_FIELD},
+			format: FORMAT_STRING, containsPII: map[string]bool{"[%START_TIME%]": true},
+			want: Out{logs: []string{VALID_LOG}, err: nil},
+		},
+
+		// negative test cases
+		"empty_logs": {logs: []string{},
+			format: FORMAT_STRING, containsPII: map[string]bool{"[%START_TIME%]": true},
+			want: Out{logs: []string{}, err: nil},
+		},
+		"empty_format_string": {logs: []string{VALID_LOG, LOG_WITH_ONE_FIELD},
+			format: "", containsPII: map[string]bool{"[%START_TIME%]": true},
+			want: Out{logs: []string{}, err: nil},
+		},
+		"empty_pii_map": {logs: []string{VALID_LOG, LOG_WITH_ONE_FIELD},
+			format: FORMAT_STRING, containsPII: map[string]bool{},
+			want: Out{logs: []string{VALID_LOG}, err: nil},
+		},
+	}
+
+	// defining subtest with t.Run
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			actualLogs, actualErr := ProcessLogs(tc.logs, tc.format, tc.containsPII)
+
+			// hash the desired fields for each log in wanted logs
```

Comment:
I have a question about what's the best way to define the expected output for my test cases in the table. Instead of defining the desired output of my test cases as hashed values, I did the logic of hashing the expected logs here. May I know if this may cause any problems? Is there a better way to do it? The obvious way is to put the hash of the relevant PII fields altogether but I feel that it would be too much manual work if there are many test cases.

Reply:
Yes, this is roughly the correct way to approach this test with the code as-written. However, this is _exactly_ the kind of test you want to avoid, as the test logic is as complex as (in fact, nearly identical to) the code under test.

I planned on us running into exactly this problem 🙂. There are two major things we'll end up changing with this code before we merge it, and this is one of them. Ability to test is the main reason I mentioned abstracting out `hash` in earlier comments. The answer here is that we should extract out the function that `ProcessLogs` calls on each piece of PII and pass that function in as a parameter to `ProcessLogs` (i.e. `actual := ProcessLogs(logs, format, containsPII)` to `actual := ProcessLogs(logs, format, containsPII, hash)`; remember that go has first class functions).

Then in our tests, we can substitute in some easy function, for example `function(piiData string) string { return "pii" }` or something similar. We can easily take a log line for a test case and substitute `pii` for the appropriate piece by hand.

Then your test should be as simple as:
```go
piiFunc := func(string) string { return "pii" }

for name, c := range tests {
	t.Run(name, func(t *testing.T) {
		actual, err := ProcessLogs(c.in, c.format, c.pii, piiFunc)
		if err != nil {
			// got an error, see if it was expected and if so if it matches
			if c.err == "" {
				t.Fatalf("ProcessLogs(%v, %v, %v, piiFunc) = %v, %v, expected no error", c.in, c.format, c.pii, actual, err)
			} else if !strings.Contains(err.Error(), c.err) {
				t.Fatalf("ProcessLogs(%v, %v, %v, piiFunc) = %v, %v, but expected err: %v", c.in, c.format, c.pii, actual, err, c.err)
			}
		} else if case.err != "" {
			// expected an error but didn't get one
			t.Fatalf("ProcessLogs(%v, %v, %v, piiFunc) = %v, %v, but expected err: %v", c.in, c.format, c.pii, actual, err, c.err)
		}
		if !reflect.DeepEqual(actual, c.out) {
			t.Fatalf("ProcessLogs(%v, %v, %v, piiFunc) = %v but expected output: %v", c.in, c.format, c.pii, actual, c.out)
		}
	}
}
```
(You can do some boolean algebra to simplify that block of `if` statements if you'd like - this is a common pattern you'll repeat in _many_ tests you write. I might even suggest saving a snippet in your IDE that creates a test skeleton with this core `for name, c := range cases` loop + the declaration of `cases := map[string]string{}{}`.)